orchagent/neighorch.cpp

```diff
@@ -1133,8 +1133,30 @@ void NeighOrch::doVoqSystemNeighTask(Consumer &consumer)
             }
 
             if (m_syncdNeighbors.find(neighbor_entry) == m_syncdNeighbors.end() ||
-                    m_syncdNeighbors[neighbor_entry].mac != mac_address)
+                    m_syncdNeighbors[neighbor_entry].mac != mac_address ||
+                    m_syncdNeighbors[neighbor_entry].voq_encap_index != encap_index)
             {
+                // Handle encap index change. SAI does not support change of encap index for
+                // existing neighbors. Remove the neighbor but do not errase from consumer sync
+                // buffer. The next iteration will add the neighbor back with new encap index
+
+                if (m_syncdNeighbors.find(neighbor_entry) != m_syncdNeighbors.end() &&
+                    m_syncdNeighbors[neighbor_entry].voq_encap_index != encap_index)
+                {
+                    //Remove neigh from SAI
+                    if (removeNeighbor(neighbor_entry))
+                    {
+                        //neigh successfully deleted from SAI. Set STATE DB to signal to remove entries from kernel
+                        m_stateSystemNeighTable->del(state_key);
```

Comment:
Why we only do this when removeNeighbor succeeds?  If removeNeighbor fails for some reasons, kernel entries are not removed and then traffic may still be forwarded with wrong encap index.  Should we signal removing kernel entries always when encap index changes?

Reply:
We do this for neat re-trying. Removal of kernel neigh without successful removal of neigh from SAI is partial processing. This will create problems in retries without additional logic.